python/ray/serve/_private/serve_request_response.py

```diff
@@ -0,0 +1,152 @@
+
+    @property
+    def client(self) -> str:
+        return self.scope.get("client", "")
+
+    @property
+    def method(self) -> str:
+        return self.scope.get("method", "websocket").upper()
+
+    @property
+    def root_path(self) -> str:
+        return self.scope.get("root_path", "")
+
+    @property
+    def route_path(self) -> str:
+        return self.scope.get("path", "")[len(self.root_path) :]
+
+    @property
+    def path(self) -> str:
+        return self.scope.get("path", "")
+
+    @property
+    def headers(self) -> List[Tuple[bytes, bytes]]:
+        return self.scope.get("headers", [])
+
+    def set_path(self, path: str):
+        self.scope["path"] = path
+
+    def set_root_path(self, root_path: str):
+        self.scope["root_path"] = root_path
+
+    def request_object(self, proxy_handle) -> StreamingHTTPRequest:
+        return StreamingHTTPRequest(
+            pickled_asgi_scope=pickle.dumps(self.scope),
+            http_proxy_handle=proxy_handle,
+        )
+
+
+class gRPCServeRequest(ServeRequest):
+    """ServeRequest implementation to wrap gRPC request protobuf and metadata."""
+
+    def __init__(
+        self,
+        request_proto: Any,
+        context: "grpc._cython.cygrpc._ServicerContext",
+        match_target: Callable[[str], Optional[str]],
+        service_method: str,
+        stream: bool,
+    ):
+        self.request = request_proto
+        self.context = context
+        self.match_target = match_target
+        self.service_method = service_method
+        self.stream = stream
+        self.app_name = ""
+        self.route_path = ""
+        self.request_id = None
+        self.method_name = "__call__"
+        self.multiplexed_model_id = DEFAULT.VALUE
+        self.setup_variables()
+
+    def setup_variables(self):
+        if self.service_method == "/ray.serve.RayServeAPIService/ListApplications":
+            self.route_path = "/-/routes"
+        elif self.service_method == "/ray.serve.RayServeAPIService/Healthz":
+            self.route_path = "/-/healthz"
+        else:
+            service_method_split = self.service_method.split("/")
+            self.request = pickle.dumps(self.request)
+            self.method_name = service_method_split[-1].lower()
+            for key, value in self.context.invocation_metadata():
+                if key == "application":
+                    self.app_name = value
+                elif key == "request_id":
+                    self.request_id = value
+                elif key == "multiplexed_model_id":
+                    self.multiplexed_model_id = value
+            self.route_path = self.match_target(self.app_name) or ""
+
+    @property
+    def user_request(self) -> bytes:
+        return self.request
+
+    @property
+    def request_type(self) -> str:
+        return "grpc"
+
+    @property
+    def method(self) -> str:
+        return "GRPC"
+
+    def send_request_id(self, request_id: str):
```

Comment:
nit: set_request_id? same for the following func name.

Reply:
because in the implementation, it is using `set_trailing_metadata([("request_id", request_id)])`. I think the function is used for setting request id information into the training metadata.

Reply:
Understood, but still feel it will cause confusing to call `set_xxx` on a model like this. Will leave as is if there is no strong opinion on this.